doc/dev/kdb-operations.md

```diff
@@ -133,53 +133,47 @@ The purpose of the `set` operation is to write data from a `KDB` instance into b
 
 Properties of `kdbSet()`:
 
+<!-- TODO: what about spec/copy? Should that really not be reflected? With the steps below the changes would show up... -->
```

Comment:
input welcome

Reply:
Please shortly describe why the changes would show up.

Reply:
```c
Key * parentKey = keyNew ("/foo", KEY_END);
kdbGet (kdb, ks, parentKey);
Key * x = keyNew ("user:/foo/new/key", KEY_END);
ksAppendKey (ks, x);
kdbSet (kdb, ks, parentKey);
```

If there is a `spec:/foo/new/key` that spec must of course be applied to `user:/foo/new/key` at the start of `kdbSet` to ensure `x` is validated correctly. The question is whether the actual key `x` should be modified, or whether `kdbSet` should not touch `ks` (and it's contents) at all and the spec data would only show up after the next `kdbGet`?

Reply:
Actually it sounds like a very good idea that meta-data gets copied and the defaults get created in `elektraKdbSet` **if** the spec implementation doesn't get too complicated.

The alternative would be that we require applications to always use `elektraKdbGet` afterwards. And this would get quite complicated, as in the error case applications already need to call `elektraKdbGet` and then merge the ks.

```c
ret = elektraKdbSet(kdb, ks...);
if (ret == -1) {
    elektraKdbGet(kdb, ksdup...);
    // merge ks and ksdup to ks
   elektraKdbSet(kdb, ks, ...)
} else if (ret == 1) {
  // also on success we need elektraKdbGet but not if nothing was changed ret==0?
  elektraKdbGet(kdb, ks...);
}
// now "ks" is allowed to be used again by the application
```

IMHO examples/kdbset.c is already way too complicated!

Reply:
> the spec implementation doesn't get too complicated.

If you look at the operational sequences I wrote for `get` and `set`, both use the hook `spec/copy`. That was intentional, because AFAICT the exact same code could be used both `get` and `set`. In `set` just need an additional `spec/remove` hook called a bit later, to remove the `meta:/` keys before we store the config.

The question here is actually just, whether `spec/copy` should be called

1. before `ksDeepDup` and therefore make the updated `meta:/` keys visible outside `kdbSet`
2. or after `ksDeepDup` thereby keeping all `spec` related changes local to `kdbSet`

Reply:
It sounds to me like a feature if the spec related changes are visible in the KeySet the application has. Then applications can simply fire off "elektraKdbSet" and, if they don't care if it is actually persisted, are done (The ks they have is consistent and complete). Or do I miss here something important?

Reply:
> if they don't care if it is actually persisted

Well they should also know that, because it should be persisted if `kdbSet` doesn't return an error. But of course there could (in theory) be things done by (less well designed) plugins that only show up after a `kdbGet`.

> Or do I miss here something important?

No, I also see this a feature for `kdbSet`.